### PR TITLE
CORTX-29861: Refactor solution.yaml to support multiple data pods per node

### PR DESCRIPTION
**Pull request by osowski**
## Description

As part of CORTX-29861, multiple structural changes are required to go into the deployment of CORTX on Kubernetes. This PR is one of two that will be required as part of 29861 and introduces the changes to the underlying solution.yaml structure and associated scripts. This is required as allow more natural definition of elements inside the solution.yaml structure with proper YAML types, as well as setting up a natural evolution towards Kubernetes Operator Custom Resource Definitions.

## Breaking change

This is a breaking change to the structure of solution.yaml and its parsing by the deploy-cortx-cloud.sh script. All necessary structural updates have been included in the updated `solution.example.yaml`. There is only one net-new parameter, but the largest changes are the restructuring of the `nodes` and `storage` sections under a new elevated (and properly made into a YAML list)  `storage_sets` section.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds new functionality)
- [x] Breaking change (bug fix or new feature that breaks existing functionality)
- [ ] Third-party dependency update
- [x] Documentation additions or improvements
- [x] Code quality improvements to existing code or test additions/updates

## Applicable issues

- This change fixes an issue: #CORTX-29861
- This change is related to an issue: #

## CORTX image version requirements

This change requires the following images:

- `cortx-data:825`
- `cortx-rgw:825`
- `cortx-control:825`

## How was this tested?

Locally with multiple deploy/destroy loops, `hctl status` reports, and basic S3 IO.

## Checklist

- [x] The change is tested and works locally.
- [x] New or changed settings in the solution YAML are documented clearly in the README.md file.
- [x] All commits are signed off and are in agreement with the [CORTX Community DCO and CLA policy](https://github.com/Seagate/cortx/blob/main/doc/dco_cla.md).

If this change requires newer CORTX or third party image versions:

- [ ] The `image` fields in [solution.example.yaml](../k8_cortx_cloud/solution.example.yaml) have been updated to use the required versions.
- [ ] The `appVersion` field of the [Helm chart](../charts/cortx/Chart.yaml) has been updated to use the new CORTX version.

If this change addresses a CORTX Jira issue:

- [x] The title of the PR starts with the issue ID (e.g. `CORTX-XXXXX:`)

-----
[View rendered README.md](https://github.com/Seagate/cortx-k8s/blob/CORTX-29861_refactor_solution_yaml/README.md)

### CODE REVIEW

File: k8_cortx_cloud/prereq-deploy-cortx-cloud.sh [hunk 1]
```diff
@@ -348,24 +347,20 @@ function symlinkBlockDevices()
...
     rm -f "${job_file}"
 
     # Iterate over the defined nodes in solution.yaml
-    filter="solution.nodes.node*.name"
-    node_output=$(parseSolution "${solution_yaml}" "${filter}")
-    IFS=';' read -r -a node_array <<< "${node_output}"
+    node_output=$(yq e '.solution.storage_sets[0].nodes' -o=c "${solution_yaml}")
+
+    # Split parsed output into an array
+    IFS=',' read -r -a node_array <<< "${node_output}"
```
Comment (keithpine) [inline k8_cortx_cloud/prereq-deploy-cortx-cloud.sh:362]:
```suggestion
    IFS=',' read -r -a node_array < <(yq ...)
```
Could also get rid of the temporary variable `node_output`.

File: k8_cortx_cloud/status-cortx-cloud.sh [hunk 2]
```diff
@@ -57,8 +57,9 @@ readonly release_selector="app.kubernetes.io/instance=cortx"
 readonly cortx_selector="${release_selector},app.kubernetes.io/name=cortx"
 
 namespace=$(parseSolution 'solution.namespace' | cut -f2 -d'>')
-num_nodes=$(parseSolution 'solution.nodes.node*.name' | grep -o '>' | wc -l)
-num_devices=$(parseSolution 'solution.storage.cvg*.devices*.device' | grep -o '>' | wc -l)
+num_nodes=$(yq '.solution.storage_sets[0].nodes | length' "${solution_yaml}")
+num_devices=$(yq '[.solution.storage_sets[0].storage[].devices.metadata.device, 
+                    .solution.storage_sets[0].storage[].devices.data[].device] | length' "${solution_yaml}")
```
Comment (keithpine) [inline k8_cortx_cloud/status-cortx-cloud.sh:62]:
```suggestion
num_devices=$(yq '[.solution.storage_sets[0].storage[].devices.metadata.device,
                   .solution.storage_sets[0].storage[].devices.data[].device] | length' "${solution_yaml}")
```

File: README.md [hunk 3]
```diff
@@ -311,29 +307,33 @@ This section contains common parameters that affect all CORTX components running
...
-| `nodes.node1.name`       | Kubernetes node name for the first node in the Kubernetes cluster available to deploy CORTX components.  | `node-1`                |
-| `nodes.node2.name`       | Kubernetes node name for the second node in the Kubernetes cluster available to deploy CORTX components. | `node-2`                |
-| `nodes.node{N}.name`     | Kubernetes node name for the Nth node in the Kubernetes cluster available to deploy CORTX components.    | None                    |
+| Name                     | Description                                                                                      | Default Value           |
+| ------------------------ | ------------------------------------------------------------------------------------------------ | ----------------------- |
+| `storage_sets`           | A list of the storage defined for use by CORTX. At this time, only one storage set is supported. | See `solution.example.yaml`  |
+| `storage_sets[].name`    | The name of an indvidual storage set.    | `storage-set-1`         |
```
Comment (keithpine) [inline README.md:323]:
```suggestion
| `storage_sets[].name`    | The name of an individual storage set.    | `storage-set-1`         |
```

File: README.md [hunk 4]
```diff
@@ -311,29 +307,33 @@ This section contains common parameters that affect all CORTX components running
...
+| `storage_sets[].durability.sns` | `TBD`                               | `1+0+0` |
+| `storage_sets[].durability.dix` | `TBD`                               | `1+0+0` |
+| `storage_sets[].container_group_size` | This value determines the number of Motr IO containers inside of a single CORTX Data Pod. This value can be tuned for optimal performance based upon different Kubernetes environments. | `1` |
```
Comment (keithpine) [inline README.md:326]:
```suggestion
| `storage_sets[].container_group_size` | This value determines the number of Motr IO containers inside of a single CORTX Data Pod. This value can be tuned for optimal performance based upon different Kubernetes environments. | `1` |
```

File: k8_cortx_cloud/deploy-cortx-cloud.sh [hunk 5]
```diff
@@ -476,20 +472,20 @@ printf "\n"
 
 namespace=$(parseSolution 'solution.namespace')
 namespace=$(echo "${namespace}" | cut -f2 -d'>')
-parsed_node_output=$(parseSolution 'solution.nodes.node*.name')
+### TODO CORTX-29861 Revisit for best way to parse this YAML section with new schema references
+parsed_node_output=$(yq e '.solution.storage_sets[0].nodes' -o=c "${solution_yaml}")
 
 # Split parsed output into an array of vars and vals
-IFS=';' read -r -a parsed_var_val_array <<< "${parsed_node_output}"
+IFS=',' read -r -a parsed_node_array <<< "${parsed_node_output}"
```
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:479]:
```suggestion
IFS=',' read -r -a parsed_node_array < <(yq '.solution.storage_sets[0].nodes' --output-format=csv "${solution_yaml}")
```
Can also remove the need for the temporary string.

**Review by keithpine — APPROVED** [review]
LGTM. All I have are a bunch of nitpick comments, you are free to ignore them if you want.

**Review by walterlopatka — APPROVED** [review]
lgtm

**Pull request merged**